runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/Job.java

```diff
@@ -654,6 +732,72 @@ public final void schedule() {
 		super.schedule(0L);
 	}
 
+	/**
+	 * Schedule the job and returns a {@link CompletableFuture} that can be used to
+	 * observe the job state:
+	 * <ul>
+	 * <li>If the job is canceled, the future will be canceled and vice versa</li>
+	 * <li>When the job completes with an CANCEL status the future will be
+	 * canceled</li>
+	 * <li>When the job completes with an ERROR status the future will complete
+	 * exceptionally with a {@link CoreException}</li>
+	 * <li>In all other cases the future is completed with the job execution
+	 * status.</li>
+	 * </ul>
+	 *
+	 * @see #schedule()
+	 * @since 3.14
+	 */
+	public final CompletableFuture<IStatus> future() {
+		CompletableFuture<IStatus> future = new CompletableFuture<>();
+		JobChangeAdapter listener = new JobChangeAdapter() {
+
+			volatile boolean sceduled;
+
+			@Override
+			public void done(IJobChangeEvent event) {
+				if (sceduled && event.getJob() == Job.this) {
+					removeJobChangeListener(this);
+					if (future.isDone()) {
+						// nothing more to do...
+						return;
+					}
+					IStatus result = event.getResult();
+					if (result.getSeverity() == IStatus.CANCEL) {
+						future.cancel(true);
+						return;
+					}
+					if (result.getSeverity() == IStatus.ERROR) {
+						future.completeExceptionally(new CoreException(result));
+						return;
+					}
+					future.complete(result);
+				}
+			}
+
+			@Override
+			public void scheduled(IJobChangeEvent event) {
+				if (event.getJob() == Job.this) {
+					sceduled = true;
+				}
+			}
+		};
+		addJobChangeListener(listener);
+		if (trySchedule(0)) {
+			future.exceptionally(ex -> {
+				if (ex instanceof CancellationException) {
+					cancel();
```

Comment:
This might cancel a subsequently scheduled run of the same job.

- Job created
- Future obtained
- Job running
- Job being cancelled via Job.cancel
- Job being rescheduled via Job.schedule
- Running Job checks the progress monitors and does indeed cancel now
- Listener is notified, future is cancelled
- Rescheduled Job is now being cancelled from here

Again this *might* not be a real problem but *if* someone is bitten by this, it'll be very hard to find, understand, and fix.

Reply:
Actually there is no guarantee today that a cancel cancel "that" execution, it just cancels whatever is currently running, so we sadly can't do anything better here. So cancel7rescedule is already hard to debug today, one reason for that I best would always call `Job.create().future()` and leave out of band cancel/schedule alone :-)